src/shim/shim.c

```diff
@@ -298,6 +302,125 @@ static void _shim_parent_init_ptrace() {
     }
 }
 
+static void _handle_sigsys(int sig, siginfo_t* info, void* voidUcontext) {
+    ucontext_t* ctx = (ucontext_t*)(voidUcontext);
+    if (sig != SIGSYS) {
+        abort();
+    }
+    // Make the syscall via the *the shim's* syscall function (which overrides
+    // libc's).  It in turn will either emulate it or (if interposition is
+    // disabled), or make the call natively. In the latter case, the syscall
+    // will be permitted to execute by the seccomp filter.
+    long rv = shadow_raw_syscall(ctx->uc_mcontext.gregs[REG_RAX], ctx->uc_mcontext.gregs[REG_RDI],
+                                 ctx->uc_mcontext.gregs[REG_RSI], ctx->uc_mcontext.gregs[REG_RDX],
+                                 ctx->uc_mcontext.gregs[REG_R10], ctx->uc_mcontext.gregs[REG_R8],
+                                 ctx->uc_mcontext.gregs[REG_R9]);
+    ctx->uc_mcontext.gregs[REG_RAX] = rv;
+}
+
+static void _shim_parent_init_seccomp() {
+    // Install signal sigsys signal handler, which will receive syscalls that
+    // get stopped by the seccomp filter.
+    if (sigaction(SIGSYS,
+                  &(struct sigaction){
+                      .sa_sigaction = _handle_sigsys,
+                      // SA_NODEFER: Allow recursive signal handling, to handle a syscall
+                      // being made during the handling of another.
+                      // SA_SIGINFO: Required because we're specifying sa_sigaction.
+                      .sa_flags = SA_NODEFER | SA_SIGINFO,
+                  },
+                  NULL) < 0) {
+        panic("sigaction: %s", strerror(errno));
+    }
+    sigset_t mask;
+    sigemptyset(&mask);
+    sigaddset(&mask, SIGSYS);
+    if (sigprocmask(SIG_UNBLOCK, &mask, NULL)) {
+        panic("sigprocmask: %s", strerror(errno));
+    }
```

Comment:
Signal masks are per-thread I believe. Should this be run once for each thread?

Reply:
This code runs in the first thread, and the mask is inherited by new threads. Added a comment.